Migrate the server-agent communication from RMI to WebSocket

dim4o opened the pull request:
Update the Agent's database scheme
Add a WebSocket layer to the Server 
Apply the WebSocket layer to the project
Fix Server's tests

dim4o committed:
Update the Agent's database scheme

Remove the ip and hostname rows from the Agent table
Remove the ip and hostname properties from the Dao and update the dependent classes

dim4o committed:
Add WebSocket layer to the Server

dim4o committed:
Fix Server's tests

Perseverance commented on src/main/java/com/musala/atmosphere/server/websocket/ServerWebsocketDispatcher.java:
```diff
@@ -0,0 +1,266 @@
+     *         - if the connection has been closed from the Agent
+     * @throws IOException
+     */
+    public void sendPing(String agentId) throws IllegalStateException, IOException {
```
Rename this to sendPong and add method sendPing that sends Ping

dim4o replied:
done

Perseverance commented on src/main/java/com/musala/atmosphere/server/Server.java:
```diff
@@ -57,6 +61,8 @@
 
     private boolean isConnected;
 
+    private ServerWebsocketDispatcher despatcher = ServerWebsocketDispatcher.getInstance();
```
Dispatcher instead of despatcher

dim4o replied:
done

Perseverance commented on src/main/java/com/musala/atmosphere/server/Server.java:
```diff
@@ -78,6 +84,9 @@ public Server() throws RemoteException {
     public Server(int serverPort) throws RemoteException {
         serverRmiPort = serverPort;
```
I don't think the RMI thing is correct

dim4o replied:
done

Perseverance commented on src/main/java/com/musala/atmosphere/server/ServerManager.java:
```diff
@@ -38,30 +37,24 @@
+     * @param agentId
      *        - the ID of the Agent
      * @param changedDeviceRmiId
      *        - the unique RMI identifier of the device
```
Please fix the RMI javadocs in the whole class

dim4o replied:
done

Perseverance commented on src/main/java/com/musala/atmosphere/server/monitor/AgentMonitor.java:
```diff
@@ -43,16 +39,15 @@ public void inform(AgentDisconnectedEvent event) {
      * @param event
      *        - event, which is received when an agent is connected.
      */
     public void inform(AgentConnectedEvent event) {
```
Let's put the logic for connect before the logic for disconnect

dim4o replied:
done

dim4o committed:
Apply the WebSocket layer to the project

bstrandjev commented on src/main/java/com/musala/atmosphere/server/websocket/ServerDispatcher.java:
```diff
@@ -0,0 +1,353 @@
+        } catch (CommandFailedException ex) {
+            Session agentSession = agentIdToAgentSessionCache.get(agentId);
+            sendErrorResponseMessage(ex, agentSession, deviceListChangedRequest.getSessionId());
+            ex.printStackTrace();
```
be consistent to your approach to exceptions - sometimes you print stack trace, sometimes you use LOGGER. I believe the latter is better

dim4o replied:
Done.

bstrandjev commented on src/main/java/com/musala/atmosphere/server/websocket/ServerDispatcher.java:
```diff
@@ -0,0 +1,353 @@
+     * @param json
+     *        - JSON message
+     */
+    void sendToClient(String json) {
```
please verify this really needs to be package protected, not private

dim4o replied:
Called from another class in the same package: `ServerAgentEndpoint`.

bstrandjev commented on src/main/java/com/musala/atmosphere/server/websocket/ServerDispatcher.java:
```diff
@@ -0,0 +1,353 @@
+     * @throws NoAvailableDeviceFoundException
+     *         - if a device was found but is currently allocated by another client
+     */
+    void sendGetDeviceAllocationInfoRequest(RequestMessage getDeviceAllocationInformationRequest,
```
verify package protected

dim4o replied:
Called from another class in the same package: `ClientServerEndpoint`.

bstrandjev commented on src/main/java/com/musala/atmosphere/server/websocket/ServerDispatcher.java:
```diff
@@ -0,0 +1,353 @@
+    /**
+     * Returns {@link DeviceAllocationInformation} of a device for the provided selector.
+     *
+     * @param deviceSelector
```
this java doc is not for this method

dim4o replied:
Done.

bstrandjev commented on src/main/java/com/musala/atmosphere/server/websocket/ServerAgentEndpoint.java:
```diff
@@ -0,0 +1,58 @@
+
+    @OnMessage
+    public void onJsonMessage(String jsonMessage, Session session) {
+        // LOGGER.debug("Server: onMessage called");
```
please either remove or uncomment

dim4o replied:
Removed.

bstrandjev commented on src/main/java/com/musala/atmosphere/server/pool/PoolManager.java:
```diff
@@ -37,31 +29,26 @@
-
-    private static final String DEVICE_RMI_ID_FORMAT = "%s_%s";
+public class PoolManager implements Subscriber {
+    private static final String DEVICE_ID_FORMAT = "%s_%s";
```
aha, here is the format - you can reuse this in the previous file i commented on

dim4o replied:
I didn't find this comment but I think I understood this remark. `buildDeviceIdentifier(...)` can be used statically: `Poolmanager.buildDeviceIdentifier(..)` and I don't need of something like `DEVICE_ID_FORMATTER` in the `ServerManager`.

bstrandjev commented on src/main/java/com/musala/atmosphere/server/pool/PoolManager.java:
```diff
@@ -262,35 +207,34 @@ public synchronized DeviceAllocationInformation allocateDevice(DeviceSelector de
-        	serialNumberAndModelList.add(new Pair<String, String>(deviceInformation.getSerialNumber(), deviceInformation.getModel()));
+        ArrayList<Pair<String, String>> serialNumberAndModelList = new ArrayList<>();
+
+        for (int index = 0; index < availableDevicesList.size(); index++) {
```
Prefer enhanced for loops

dim4o replied:
Done.

bstrandjev reviewed (commented):
I think github might have got a bit troubled with the size of your changes - please note that there should be about 4 comments on this commit

bstrandjev reviewed (approved):
I love your concept of "Fix tests"

dim4o committed:
Code review changes (for squash)

Code review changes
Remove the rmiId from the methods and rename or remove the related properties and fields
Fix tests

dim4o committed:
Fix ConcurrentModificationException when trying to remove all devices

dim4o committed:
Fix device ID cache

The device ID should be removed from the `deviceIdCache` when the `removeDevice(deviceId)` method is called
